bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyBaseHandler.java

```diff
@@ -0,0 +1,810 @@
+    private @Nullable ShellyCoapHandler      coap;
+    protected @Nullable ShellyDeviceProfile  profile;
+    private final @Nullable ShellyCoapServer coapServer;
+
+    @SuppressWarnings("unused")
+    private long                             lastUpdateTs     = 0;
+    private long                             lastUptime       = 0;
+    private long                             lastAlarmTs      = 0;
+
+    private @Nullable ScheduledFuture<?>     statusJob;
+    private int                              skipUpdate       = 0;
+    public int                               scheduledUpdates = 0;
+    private int                              skipCount        = UPDATE_SKIP_COUNT;
+    private int                              refreshCount     = UPDATE_SETTINGS_INTERVAL_SECONDS
+            / UPDATE_STATUS_INTERVAL_SECONDS;                                                                    // force
+                                                                                                                 // settings
+                                                                                                                 // refresh
+                                                                                                                 // every
+                                                                                                                 // x
+    // seconds
+    private final int                        cacheCount       = UPDATE_SETTINGS_INTERVAL_SECONDS
+            / UPDATE_STATUS_INTERVAL_SECONDS;                                                                    // delay
+                                                                                                                 // before
+                                                                                                                 // enabling
+                                                                                                                 // channel
+    // cache
+    private boolean                          refreshSettings  = false;
+    private boolean                          channelCache     = false;
+    protected boolean                        lockUpdates      = false;
+
+    public String                            thingName        = "";
+    private Map<String, Object>              channelData      = new HashMap<>();
+    protected ShellyBindingConfiguration     bindingConfig    = new ShellyBindingConfiguration();
+
+    /**
+     *
+     * @param thing The Thing object
+     * @param handlerFactory Handler Factory instance (will be used for event
+     *            handler registration)
+     * @param bindingConfig The binding configuration (beside thing
+     *            configuration)
+     * @param networkAddressService Network service to get local ip
+     */
+    public ShellyBaseHandler(Thing thing, ShellyHandlerFactory handlerFactory, ShellyBindingConfiguration bindingConfig,
```

Comment:
This requires refactoring. The ThingHandler must not depend on the handler factory - so whatever you need it for here, think of a better way to achieve it and possibly pass more specific objects through the constructor.

Reply:
I would like to push this out after the release with the given timeline - is that ok for you?
And maybe need some help.

Reply:
No, this one really must be addressed before the release. I'll assist you if there's an issue with how the refactoring could be done.

Reply:
yes please

Reply:
Ok, so can you point me to the places that are problematic?
- I can see you use `ShellyHandlerFactory.convertTimestamp`, which is merely a static method that you can move to some utility class without issues.
- You also call `registerDeviceListener(this)`. You could e.g. do this in the `ThingHandlerFactory. createHandler` already - you might then only need to make sure that the `handler.onEvent` implementation checks whether the handler is already successfully initialized before processing the event.

Anything more to address?

Reply:
done, that was easy, thx